tools/common/schema/updatetask.go

```diff
@@ -392,8 +428,22 @@ func sortAndFilterVersions(versions []string, startVerExcl string, endVerIncl st
 // the schema changes for versions in the range startVer < ver <= endVer
 // when endVer is empty this method returns all subdir names that are greater than startVer
 func readSchemaDir(dir string, startVer string, endVer string, logger log.Logger) ([]string, error) {
+	return readSchemaDirEFS(dir, startVer, endVer, logger, nil)
+}
 
-	subDirs, err := os.ReadDir(dir)
+// readSchemaDirEFS returns a sorted list of subdir names that hold
+// the schema changes for versions in the range startVer < ver <= endVer
+// when endVer is empty this method returns all subdir names that are greater than startVer.
+// Compatible with an embedded file system generated by go:embed, calling efs.ReadDir
+// but if efs == nil it uses os.ReadDir
+func readSchemaDirEFS(dir string, startVer string, endVer string, logger log.Logger, efs *embed.FS) ([]string, error) {
```

Comment:
You could simplify this by taking in an `fs.FS` instead of an `embed.FS`. You can open the "real" filesystem dir as an `FS` by using `os.DirFS(dir)` which would let you unify the code paths

Reply:
You may need to take more than just `fs.FS` to do everything you want (I didn't dig through all the calls you made), but `fs.ReadFileFS` and `fs.ReadDirFS` look to have what you need

Reply:
i did consider using an `fs.FS` + `os.DirFS(dir)` and actually wrote a bunch of code doing that that I then deleted it lol. I absolutely agree the code would be cleaner (no `if efs != nil`) if I did that, but I wasn't sure what to pass in as the `dir` for `os.DirFS` that would be sure to not break any customer's existing filesystem structures for where they are using this tool and storing their schemas.

I chose this route because I felt like it was less risky, given that I can't test for all the edge cases that might exist in our customer's file systems. 

this https://github.com/golang/go/issues/44279 made me concerned about `os.DirFS`, in addition to the note here https://pkg.go.dev/os#DirFS about limitations. if you have a sec to look at `os.DirFS` with me and talk through it together i would be super down to consider!

Reply:
Did you try it out on Windows and encounter that bug? It looks like the problem happens when using the filesystem's root. I wonder if it'll have an issue if, say, a user were to specify a path to a different filesystem like `D:\\Foo\Bar\Baz\schemas`.

I'd definitely prefer simpler code; this is an administrator tool and I highly doubt we support schemas that aren't our official ones, so the common usage will just be `temporal-cassandra-tool update-schemas relative-path-to-schema`.

Reply:
got it working with fs.FS